shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -173,6 +188,19 @@ abstract class PolyTypeImpl extends Located { self: PolyType =>
   def show: Str = s"${targs.iterator.map(_.name).mkString("[", ", ", "] ->")} ${body.show}"
 }
 
+trait TypeVarImpl extends Ordered[TypeVar] { self: TypeVar =>
+  // define an ordering for type variables
+  // order by
+  // 1. name hint - default ""
+  // 2. identifier number - default 0
+  // 3. identifier string - default ""
+  def toTuple: (String, Int, String) = self.nameHint.map((_, 0, ""))
```

Comment:
This is still wrong. You're mapping two `TypeVar` that happen to have the same name hint to _the same tuple_, although they are not the same variable! Name hints are just that – hints – they don't identify a type variable.

What I meant by "use tuples" was just to write `(identifier, nameHint) compare (that.identifier, that.nameHint)` below. Please do that and remove this meaningless `toTuple` method.

Reply:
:grimacing: fixed

Reply:
You see why this makes a difference and why your implementation was wrong, right?

Reply:
Well I ~~fixed~~ changed it but it didn't compile it (more :grimacing:). It turns out that Either type does not implement a compare, I think that's why I had (incorrectly) implemented the previous tupling thing.

Reply:
About the wrong implementation, yes I understand now that nameHint is a field but it's not to be used to uniquely identify or compare to TypeVar's.

Since now it's about comparing identifier which can be either an int or a string. Something like this should work -

```scala
    (this.identifier.fold((_, ""), (0, _))) compare (that.identifier.fold((_, ""), (0, _)))
```

Reply:
I'd be honoured, if you would attest my comprehension of the problem, the proposed solution thereof, and hence sanction consolidation of the code enclosed herein.

Reply:
Do you mean you _request_ me to do that? 😛 

Yes, your new solution seems good enough. I'm always surprised the `Either` type doesn't have an ordering instance in the standard library, when it's pretty obvious how that can be implemented. (The fact it does not have one is why my suggested code did not compiler.)